chalk-engine/src/logic.rs

```diff
@@ -222,9 +222,9 @@ impl<C: Context> Forest<C> {
             goal
         );
         let coinductive_goal = context.is_coinductive(&goal);
-        let table = self.tables.insert(goal, coinductive_goal);
-        self.push_initial_strands(context, table);
-        table
+        let mut table = Table::new(goal.clone(), coinductive_goal);
+        Self::push_initial_strands(context, self.tables.next_index(), &mut table);
```

Comment:
The other way to do this would be to have `push_initial_strands` just be `get_initial_strands` and make `Table::new` take a `Vec` of `Strand`s?

Hmm, maybe not, since it also needs to be able to mark as floundered.

Reply:
`push_intial_strands_instantiated` could just return a `Table`? Then the `mut` is only in that function.

Reply:
I prefer to rename the function to somthing like `build_table` (I find `build_table_with_initial_strands` too long, but I would doc-comment it, that it just pushes table with initial strands). Obviously function would return finalized table - I just assumed that you have some reason I am not aware to go with it this way (like being more conformant with some book you are basing on, so it is easier to follow code reading documentation), that's way I was cautious with bigger changes around.

Reply:
`build_table` works :)

In general, some *ideas* of SLG are based on the paper. But the code itself it's a "new" implementation. (In fact, since I've started with Chalk, I've refactored quite a bit of this code)